README
SciDAC QLA C interface -- Level 1 linear algebra routines

This package uses the standard GNU configuration script.  See the file
INSTALL for datails on compiling and installing.  It is recommended that
QLA be compiled with a C99 compliant compiler (or at least one that
understands the restrict keyword).  Check your compiler documentation to
find the appropriate option (for gcc try -std=c99 or for xlc try
-qlanglvl=stdc99).

**** BG/L config notes ****

The recommended configure line for BG/L is

./configure \
  --prefix=$INSTALL_DIR \
  --enable-nc=3 \
  --enable-440d \
  --enable-xlc \
  --enable-temp-precision=D \
  --enable-backend-test-cmd="$RUN_COMMAND" \
  --host=blrts --build=none \
  CC=blrts_c99 \
  CFLAGS="-O3 -qarch=440d" \
  TEST_CFLAGS="-O2"

Where $INSTALL_DIR is the directory you want it installed in and
$RUN_COMMAND is a command that can be used to run an executable on the
backend.  For cobalt this can simply be "cqsub -t 5 -n 1 -c 1 -m co".
If you can run mpirun directly to submit jobs on the backend then
something like "mpirun -np 1" might work depending on the setup.
Otherwise for loadleveler you need to make a script like the one
below, make sure it is executable and in your path and then pass that
as the run command.

#!/bin/sh
llsubmit - <<EOF
# @ job_type = parallel
# @ executable = /bgl/BlueLight/ppcfloor/bglsys/bin/mpirun
# @ arguments = -np 1 -connect MESH -mode CO -cwd `pwd` -exe $1
# @ wall_clock_limit = 10
# @ queue
EOF


If all goes well then you can do the usual "make && make install" then
"cd test" and do "make check".  This should start the tests running on
the backend.  When all the tests are finished you can do "make check-TESTS"
from the same directory (test) and that will give a summary of the tests.
You can also look at the various */*.result files directly.


**** below are the original notes ****

Some configuration options to be aware of are:
 --enable-sse	uses optimized SSE routines for some functions (single)
 --enable-sse2	uses optimized SSE2 routines for some functions (double)
 --enable-sse3	uses optimized SSE3 routines for some functions
 --enable-440	uses optimized 440 routines for some functions (QCDOC/BGL)
 --enable-nc=[2,3,n]	build only the 2, 3, or arbitrary (n) color libraries
			leave this out to build all libraries

The sse options are cumulative so that sse2 also includes sse and sse3
includes sse and sse2.

The sse (not sse2) code now works with both gcc and icc.  It will just
use the same compiler as used for the rest of the code.  You may have
to give it the appropriate commands (like -msse or -msse2) for it to
compile the sse code.

The sse2 code still requires gcc.  If the main compiler is gcc then it
will use that otherwise it will default to just using the 'gcc'
command found in the path.  You can specify the correct location of
the needed gcc compiler by setting the GNUCC environment variable.
This will only be used for the sse2 code.  Note however that icc can
emulate gcc in some cases and may actually be able to compile this
code fine.

There are now variables TEST_CC and TEST_CFLAGS which are used only
for compiling the test suite.  Since the test sources are very large
you may want to disable optimizations here to reduce compile time.

An example configure command is:

configure \
  --prefix=$HOME/scidac/install/qla \
  --enable-nc=3 \
  --enable-sse3 \
  CC=icc \
  CFLAGS="-c99 -O2 -xN -fno-alias -unroll -fp-model fast" \
  TEST_CFLAGS="-c99 -O0 -xN" \
  LDFLAGS="-static"

Documentation is included in the doc directory.  This is not included in
the installation.  See the README file in the doc directory for details.

Contents of directories in src

  doc		Documentation - Postscript and HTML
  examples	A simple benchmark routine for a few functions. (make check)
  include	Predefined headers and generated headers
  lib/cmath	Complex arithmetic sources
  lib/perl	Perl script for building the libraries
  lib/random	Random number utilities
  lib/c99	c99 functions for non-c99 compliant compilers.
  lib/optimized	Various optimized routines for different platforms.
  lib/qla_*	Contains makefile for each library.
  tests		Test suite.  This is run with 'make check'.


Bugs and suggestions:
James C. Osborn
Boston University
josborn at physics.bu.edu


NEWS
===============================================================================
2014-08-08	version 1.9.0
===============================================================================
2014-08-07: James Osborn
	Added local Xa_dot_X and X_dot_Xa and global Xa_dot_X routines.

2014-02-28: James Osborn
	Made testsuite run with varying alignments.
	Corrected non-aligned sse spin project/reconstruct code.

2014-01-16: James Osborn
	Made generic cmath macros follow QLA_Precision.

===============================================================================
2013-10-26	version 1.8.0
===============================================================================
2013-10-13: James Osborn
	Added virtual Nc=1 library.

2013-05-15: James Osborn
	Added precision specific versions of some macros for C99 complex.

===============================================================================
2013-05-10	version 1.7.3
===============================================================================
2013-05-10: James Osborn
	Fixed special cases of matrix functions.

===============================================================================
2013-05-04	version 1.7.2
===============================================================================
2013-05-04: James Osborn
	Added a few more matrix functions and optimized many for Nc=2,3.

2013-03-26: James Osborn
	Added more matrix and scalar fill functions.

2013-03-07: James Osborn
	Added M_inverse_V function.

===============================================================================
2013-02-01	version 1.7.1
===============================================================================
2013-02-01: James Osborn
	Added QLA_PrecisionInt macro.
	Added experimental multi-source functions.
	Added some QPX routines.
	Added precision generic headers.

===============================================================================
2012-01-11	version 1.7.0-a7 (7th alpha release of 1.7.0)
===============================================================================
2012-01-03: James Osborn
	Added float versions of cmath functions.

2011-10-19: James Osborn
	Added invsqrt_M function.
	Added alignment hints to types.

2011-10-18: James Osborn
	Added some optimized C routines for BG/P.
	Added int and real type min and max macros.

2011-09-24: James Osborn
	Changed gaussian random number routine to optionally match MILC.

===============================================================================
2011-01-22	version 1.7.0-a6 (6th alpha release of 1.7.0)
===============================================================================
2011-01-21: James Osborn
	Added OpenMP support for all routines.

2010-08-24: James Osborn
	Fixed OpenMP support.

2010-04-22: James Osborn
	Fixed mem & flops for QLA_D_vpeq_spproj_D benchmark.

===============================================================================
2010-04-22	version 1.7.0-a5 (5th alpha release of 1.7.0)
===============================================================================
2010-04-20: James Osborn
	Added more functions, removed C_eqop_conj_C, added temporary pointers
	for indexed varibles to avoid XLCv8 bug.

2010-04-09: James Osborn
	Added more times functions.

2010-04-04: James Osborn
	Fix _div_c macros to use safer temporary name.

===============================================================================
2010-04-04	version 1.7.0-a4 (4th alpha release of 1.7.0)
===============================================================================
2010-04-04: James Osborn
	Many additions:
          r_div_c macro
          QLA_version_str()
	  QLA_version_int()
	  QLA_C_eq_det_M()
	  QLA_M_eq_inverse_M()
	  QLA_M_eq_exp_M()
	  QLA_M_eq_sqrt_M()
	  QLA_M_eq_log_M()
	  QLA_X_eq_R_times_X()
	  QLA_X_eq_C_times_X()
	   (for X = V,H,D,M,P)
	  2 & N color example benchmarks

===============================================================================
2009-11-05	version 1.7.0-a3 (3rd alpha release of 1.7.0)
===============================================================================
2009-09-28: James Osborn
        Added missing qla_complex_c99.h

===============================================================================
2009-09-10	version 1.7.0-a2 (2nd alpha release of 1.7.0)
===============================================================================
2009-08-30: James Osborn
	Numerous code generation improvements and fixes.

2009-08-24: James Osborn
	Fixed bug in sse code.

===============================================================================
2009-05-20	version 1.7.0-a1 (first alpha release of 1.7.0)
===============================================================================
2009-05-02: James Osborn
	Added more configure/code generation options mainly related to BG/L&P.

2009-04-20: James Osborn
	Changed 'N' color types to be any size and only as large as needed.
	The QLA_[DF]N_... type is now a macro taking Nc and the variable
	as arguments, i.e.
	  QLA_FN_ColorMatrix(10,foo);
	declares 'foo' as a 10x10 color matrix.  To create a new type of a
	specific Nc you can do
	  typedef QLA_FN_ColorMatrix(12,CM12);
	  CM12 foo;
	  CM12 *bar = (CM12 *) malloc(nmat*sizeof(CM12));

2008-10-06: James Osborn
	Added extern "C" to qla.h.

===============================================================================
2008-09-22	version 1.6.3
===============================================================================
2008-09-22: James Osborn
	Added source-stamp to clean targets.

2008-02-18: James Osborn
	Added AC_PROG_CC_C99 to configure.ac.

2008-02-06: James Osborn
	Added OpenMP directives in some places.

2008-02-05: James Osborn
	Fixed Makefiles to allow parallel makes.

2007-11-28: James Osborn
        Fixed "expr" expression in configure.ac to work with Macs.

===============================================================================
2006-10-10	version 1.6.2
===============================================================================
2006-10-10:	James Osborn

examples/qla_bench.c
#include <stdlib.h>
#include <stdio.h>
#include <time.h>
#include <sys/time.h>
#include <math.h>
#include <qla.h>
#ifdef _OPENMP
# include <omp.h>
# define __USE_GNU
# include <sched.h>
# define PFOR _Pragma("omp parallel for")
#else
# define PFOR
#endif

#if QLA_Precision == 'F'
#define REALBYTES 4
#else
#define REALBYTES 8
#endif
#define NC QLA_Nc

// flops rdivc: 6 cdivc: 12 csqrt: 11 cpow: 11 cexp: 5 ccosh: 6
// det: 1 0, 2 14, 3 64, n(n*n-1)4/3+3n+3n(n-1)/2+6n+n(n-1)(n-2)4/3+6n(n-1)
// T(n) = n[T(n-1)+8]-2 , n[ 8.5 +10.5n +8n*n ]/3
#define FLOP_DET(n) (n==1?0:(n==2?14:(n==3?64:(n*(8.5+10.5*n+8*n*n)/3))))
// eig: 1 0, 2 56, 3 233, 16n*n*n
#define FLOP_EIG(n) (n==1?0:(n==2?56:(n==3?233:(16*n*n*n))))
// eigh: 1 0, 2 14, 3 233, 16n*n*n
#define FLOP_EIGH(n) (n==1?0:(n==2?14:(n==3?233:(16*n*n*n))))
// inv: 1 6, 2 48, 3 208, n(3(n+1)/2+3+6(n-1)+4(n-1)(n-1)+8n(n-1)+6n)
#define FLOP_INV(n) (n==1?6:(n==2?48:(n==3?208:(n*(12*n*n-2.5*(n-1))))))
// invV: 1 12, 2 60, 3 238, n(3(n+1)/2+3+6(n-1)+4(n-1)(n-1)+8(n-1)+6)
#define FLOP_INVV(n) (n==1?12:(n==2?60:(n==3?238:(n*(4*n*n+7.5*n+0.5)))))
// invM: 1 12, 2 100, 3 309(N=3), n(3(n+1)/2+3+6(n-1)+4(n-1)(n-1)+8n(n-1)+6n)
#define FLOP_INVM(n) (n==1?12:(n==2?100:(n==3?309:(n*(12*n*n-2.5*(n-1))))))
// exp: 1 5, 2 102, 3 699, (44+8(s+b))n3 + 29.5n2 + 2.5n + 3
#define FLOP_EXP(n) (n==1?5:(n==2?102:(n==3?699:(((44*NC+29.5)*NC+2.5)*NC+3))))
// expa: 1 2, 2 53, 3 561, (44+8(s+b))n3 + 29.5n2 + 2.5n + 3
#define FLOP_EXPA(n) (n==1?2:(n==2?53:(n==3?561:(((44*NC+29.5)*NC+2.5)*NC+3))))
// expta: 1 0, 2 21, 3 535, (44+8(s+b))n3 + 29.5n2 + 2.5n + 3
#define FLOP_EXPTA(n) (n==1?0:(n==2?21:(n==3?535:(((44*NC+29.5)*NC+2.5)*NC+3))))
// sqrt: 1 11, 2 124, 3 708(N=3,c=1), (20c)n3+(5.5c+12)n2+(2.5c)n+3
#define FLOP_SQRT(n) (n==1?11:(n==2?124:(n==3?708:(((40*NC+23)*NC+5)*NC+3))))
// sqrtph: 1 1, 2 24, 3 359, (20c)n3+(5.5c+12)n2+(2.5c)n+3
#define FLOP_SQRTPH(n) (n==1?1:(n==2?24:(n==3?359:(((40*NC+23)*NC+5)*NC+3))))
// rsqrt: 1 17, 2 134, 3 1018(N=3,c=1), (20c+12)*n3+(5.5c+9.5)n2+(2.5c+2.5)n+4
#define FLOP_RSQRT(n) (n==1?17:(n==2?134:(n==3?1018:(((52*NC+20.5)*NC+7.5)*NC+4))))
// rsqrtph: 1 2, 2 26, 3 365, (20c+12)*n3+(5.5c+9.5)n2+(2.5c+2.5)n+4
#define FLOP_RSQRTPH(n) (n==1?2:(n==2?26:(n==3?365:((((52*NC+20.5)*NC+7.5)*NC+4)))))
// log: 1 ?, 2 ?, 3 ??, ??
//#define FLOP_LOG(n) (n==1?5:(n==2?102:(n==3?3090:(60*NC*NC*NC))))
#define FLOP_LOG(n) (FLOP_SQRT(n)+FLOP_EXP(n))


#define myalloc(type, n) (type *) aligned_malloc(n*sizeof(type))

#define ALIGN 64
void *
aligned_malloc(size_t n)
{
  size_t m = (size_t) malloc(n+ALIGN);
  size_t r = m % ALIGN;
  if(r) m += (ALIGN - r);
  return (void *)m;
}

double
dtime(void)
{
#ifdef _OPENMP
  return omp_get_wtime();
#else
  struct timeval tv;
  gettimeofday(&tv, NULL);
  return tv.tv_sec + 1e-6*tv.tv_usec;
#endif
}

void
set_R(QLA_Real *r, int i)
{
  *r = 1+cos(i);
}

void
set_C(QLA_Complex *c, int i)
{
  QLA_c_eq_r_plus_ir(*c, 1+cos(i), 1+sin(i));
}

void
set_V(QLA_ColorVector *v, int i)
{
  for(int j=0; j<QLA_Nc; j++) {
    QLA_c_eq_r_plus_ir(QLA_elem_V(*v,j), j+1+cos(i), j+1+sin(i));
    //QLA_real(QLA_elem_V(*v,j)) = 1;
    //QLA_imag(QLA_elem_V(*v,j)) = 0;
  }
}

void
set_H(QLA_HalfFermion *h, int i)
{
  for(int j=0; j<QLA_Nc; j++) {
    for(int k=0; k<(QLA_Ns/2); k++) {
      QLA_c_eq_r_plus_ir(QLA_elem_H(*h,j,k), (j+4)*(k+1)+cos(i), (j+4)*(k+1)+sin(i));
    }
  }
}

void
set_D(QLA_DiracFermion *d, int i)
{
  for(int j=0; j<QLA_Nc; j++) {
    for(int k=0; k<QLA_Ns; k++) {
      QLA_c_eq_r_plus_ir(QLA_elem_D(*d,j,k), (j+4)*(k+1)+cos(i), (j+4)*(k+1)+sin(i));
    }
  }
}

#define MtypeH  1
#define MtypeP  2
#define MtypeA  4
#define MtypeT  8
#define MtypeNZ 16
#define MtypeNN 32
static int Mtype = 0;
#define setM Mtype = 0
#define setMH Mtype = MtypeH
#define setMPH Mtype = MtypeP|MtypeH|MtypeNN|MtypeNZ
#define setMA Mtype = MtypeA
#define setMTA Mtype = MtypeT|MtypeA
#define setMP Mtype = MtypeP|MtypeNN|MtypeNZ
#define setMNZ Mtype = MtypeNZ
#define setMNNH Mtype = MtypeNN|MtypeH

void
set_M(QLA_ColorMatrix *m, int i)
{
#if 0
  static QLA_ColorMatrix t;
  for(int j=0; j<QLA_Nc; j++) {
    for(int k=0; k<QLA_Nc; k++) {
      QLA_c_eq_r_plus_ir(QLA_elem_M(*m,j,k),
			 (((j-k+QLA_Nc+1)*(j+k+1))%19)+cos(i),
			 (((j+4)*(k+1))%17)+sin(i));
      //QLA_real(QLA_elem_M(*m,j,k)) = 1;
      //QLA_imag(QLA_elem_M(*m,j,k)) = 0;
    }
  }
#endif
  for(int j=0; j<QLA_Nc; j++) {
    for(int k=0; k<QLA_Nc; k++) {
      QLA_c_eq_r(QLA_elem_M(*m,j,k), 0);
    }
  }
  QLA_Real step = 1e-5;
  if(Mtype&MtypeNZ) {
    for(int j=0; j<QLA_Nc; j++) {
      QLA_c_peq_r_plus_ir(QLA_elem_M(*m,j,j), step, -step);
    }
  }
  int ii=i;
  if((Mtype&MtypeNN)==0) ii>>=QLA_Nc;
  for(int j=0,k=1; ii; ii>>=1,j++) {
    if(j>=QLA_Nc) { j=0; k*=2; }
    if(ii&1) QLA_c_peq_r_plus_ir(QLA_elem_M(*m,j,j), k*step, -k*step);
  }
  ii = i;
  if((Mtype&MtypeNN)==0) {
    for(int j=0; j<QLA_Nc; j++) {
      if(ii&1) QLA_c_eqm_c(QLA_elem_M(*m,j,j), QLA_elem_M(*m,j,j));
      ii >>= 1;
    }
  }
  if(Mtype&MtypeH) { // make Hermitian
    QLA_ColorMatrix m2;
    QLA_M_eq_M(&m2, m);
    QLA_M_peq_Ma(&m2, m);
    QLA_M_eq_M(m, &m2);
  }
  if((Mtype&MtypeP)&&(Mtype&MtypeH)) { // make positive Hermitian
    QLA_ColorMatrix m2;
    QLA_M_eq_M_times_Ma(&m2, m, m);
    QLA_M_eq_M(m, &m2);
  }
  if(Mtype&MtypeA) { // make anti-Hermitian
    QLA_ColorMatrix m2;
    QLA_M_eq_M(&m2, m);
    QLA_M_meq_Ma(&m2, m);
    QLA_M_eq_M(m, &m2);
  }
  if((Mtype&MtypeT)&&(Mtype&MtypeA)) { // make traceless anti-Hermitian
    QLA_ColorMatrix m2;
    QLA_M_eq_antiherm_M(&m2, m);
    QLA_M_eq_M(m, &m2);
  }
  //QLA_Real n2;
  //QLA_r_eq_norm2_M(&n2, m);
  //printf("%i\t%g\n", i, n2);
}

QLA_Real
sum_C(QLA_Complex *d, int n)
{
  QLA_Real t=0, *r=(QLA_Real *)d;
  int nn = n*sizeof(QLA_Complex)/sizeof(QLA_Real);
  for(int i=0; i<nn; i++) {
    t += r[i];
    //printf("%i\t%g\n", i, r[i]);
  }
  return t/nn;
}

QLA_Real
sum_V(QLA_ColorVector *d, int n)
{
  QLA_Real t=0, *r=(QLA_Real *)d;
  int nn = n*sizeof(QLA_ColorVector)/sizeof(QLA_Real);
  for(int i=0; i<nn; i++) t += r[i];
  return t/nn;
}

QLA_Real
sum_H(QLA_HalfFermion *d, int n)
{
  QLA_Real t=0, *r=(QLA_Real *)d;
  int nn = n*sizeof(QLA_HalfFermion)/sizeof(QLA_Real);
  for(int i=0; i<nn; i++) t += r[i];
  return t/nn;
}

QLA_Real
sum_D(QLA_DiracFermion *d, int n)
{
  QLA_Real t=0, *r=(QLA_Real *)d;
  int nn = n*sizeof(QLA_DiracFermion)/sizeof(QLA_Real);
  for(int i=0; i<nn; i++) t += r[i];
  return t/nn;
}

QLA_Real
sum_M(QLA_ColorMatrix *d, int n)
{
  QLA_Real t=0, *r=(QLA_Real *)d;
  int nn = n*sizeof(QLA_ColorMatrix)/sizeof(QLA_Real);
  for(int i=0; i<nn; i++) t += r[i];
  return t/nn;
}

#define set_fields { \
  PFOR \
  for(int i=0; i<n; ++i) { \
    set_R(&r1[i], i); \
    set_C(&c1[i], i); \
    set_V(&v1[i], i); \
    set_V(&v2[i], i); \
    set_V(&v3[i], i); \
    set_V(&v4[i], i); \
    set_V(&v5[i], i); \
    set_H(&h1[i], i); \
    set_H(&h2[i], i); \
    set_D(&d1[i], i); \
    set_D(&d2[i], i); \
    set_M(&m1[i], i); \
    set_M(&m2[i], i); \
    set_M(&m3[i], i); \
    set_M(&m4[i], i); \
    int j = ((i|16)+256) % n; \
    vp1[i] = &v2[j]; \
    vp2[i] = &v3[j]; \
    vp3[i] = &v4[j]; \
    vp4[i] = &v5[j]; \
    hp1[i] = &h2[j]; \
    dp1[i] = &d2[j]; \
    mp1[i] = &m3[j]; \
  } \
}

int
main(int argc, char *argv[])
{
  QLA_Real sum, *r1;
  QLA_Complex *c1;
  QLA_ColorVector *v1, *v2, *v3, *v4, *v5;
  QLA_ColorVector **vp1, **vp2, **vp3, **vp4;
  QLA_HalfFermion *h1, *h2, **hp1;
  QLA_DiracFermion *d1, *d2, **dp1;
  QLA_ColorMatrix *m1, *m2, *m3, *m4, **mp1;
  double cf0, flop, mem, time1;
  int nmin, nmax, c, nthreads=1;

  nmin = 64;
  if(argc>1) nmin = atoi(argv[1]);
  nmax = 256*1024;
  if(argc>2) nmax = atoi(argv[2]);
  cf0 = 1e9;
  if(argc>3) cf0 *= atof(argv[3]);

  printf("QLA version %s (%i)\n", QLA_version_str(), QLA_version_int());
  printf("QLA_Precision = %c\n", QLA_Precision);
  printf("QLA_Nc = %i\n", QLA_Nc);

#ifdef _OPENMP
  nthreads = omp_get_max_threads();
  printf("OMP threads = %i\n", nthreads);
  printf("omp_get_wtick = %g\n", omp_get_wtick());
#ifdef CPU_ZERO
#pragma omp parallel
  {
    int tid = omp_get_thread_num();
    cpu_set_t set;
    CPU_ZERO(&set);
    CPU_SET(tid, &set);
    sched_setaffinity(0, sizeof(set), &set);
  }
#endif
#endif

  nmin *= nthreads;
  nmax *= nthreads;

  r1 = myalloc(QLA_Real, nmax);
  c1 = myalloc(QLA_Complex, nmax);
  v1 = myalloc(QLA_ColorVector, nmax);
  v2 = myalloc(QLA_ColorVector, nmax);
  v3 = myalloc(QLA_ColorVector, nmax);
  v4 = myalloc(QLA_ColorVector, nmax);
  v5 = myalloc(QLA_ColorVector, nmax);
  vp1 = myalloc(QLA_ColorVector *, nmax);
  vp2 = myalloc(QLA_ColorVector *, nmax);
  vp3 = myalloc(QLA_ColorVector *, nmax);
  vp4 = myalloc(QLA_ColorVector *, nmax);
  h1 = myalloc(QLA_HalfFermion, nmax);
  h2 = myalloc(QLA_HalfFermion, nmax);
  hp1 = myalloc(QLA_HalfFermion *, nmax);
  d1 = myalloc(QLA_DiracFermion, nmax);
  d2 = myalloc(QLA_DiracFermion, nmax);
  dp1 = myalloc(QLA_DiracFermion *, nmax);
  m1 = myalloc(QLA_ColorMatrix, nmax);
  m2 = myalloc(QLA_ColorMatrix, nmax);
  m3 = myalloc(QLA_ColorMatrix, nmax);
  m4 = myalloc(QLA_ColorMatrix, nmax);
  mp1 = myalloc(QLA_ColorMatrix *, nmax);
  //QLA_ColorVector *va[4] = { v2, v3, v4, v5 };
  QLA_ColorVector **vpa[4] = { vp1, vp2, vp3, vp4 };
  QLA_ColorMatrix *ma[4] = { m1, m2, m3, m4 };

  for(int n=nmin; n<=nmax; n*=2) {
    printf("len = %i\n", n);
    printf("len/thread = %i\n", n/nthreads);
    double cf = cf0*nthreads/n;

#include "benchfuncs.c"

  }

  return 0;
}
